src/Microsoft.OData.Core/ODataBatchWriter.cs

```diff
@@ -313,81 +313,34 @@ public Task<ODataBatchOperationResponseMessage> CreateOperationResponseMessageAs
-        public Task FlushAsync()
-        {
-            this.VerifyCanFlush(false);
-
-            // make sure we switch to state FatalExceptionThrown if an exception is thrown during flushing.
-            return this.rawOutputContext.FlushAsync().FollowOnFaultWith(t => this.SetState(BatchWriterState.Error));
-        }
+        public abstract Task FlushAsync();
 #endif
 
         /// <summary>
         /// This method is called to notify that the content stream for a batch operation has been requested.
         /// </summary>
-        void IODataBatchOperationListener.BatchOperationContentStreamRequested()
-        {
-            // Write any pending data and flush the batch writer to the async buffered stream
-            this.StartBatchOperationContent();
-
-            // Flush the async buffered stream to the underlying message stream (if there's any)
-            this.rawOutputContext.FlushBuffers();
-
-            // Dispose the batch writer (since we are now writing the operation content) and set the corresponding state.
-            this.DisposeBatchWriterAndSetContentStreamRequestedState();
-        }
+        public abstract void BatchOperationContentStreamRequested();
 
 #if ODATALIB_ASYNC
         /// <summary>
         /// This method is called to notify that the content stream for a batch operation has been requested.
         /// </summary>
         /// <returns>
-        /// A task representing any action that is running as part of the status change of the operation; 
+        /// A task representing any action that is running as part of the status change of the operation;
         /// null if no such action exists.
         /// </returns>
-        Task IODataBatchOperationListener.BatchOperationContentStreamRequestedAsync()
-        {
-            // Write any pending data and flush the batch writer to the async buffered stream
-            this.StartBatchOperationContent();
-
-            // Asynchronously flush the async buffered stream to the underlying message stream (if there's any);
-            // then dispose the batch writer (since we are now writing the operation content) and set the corresponding state.
-            return this.rawOutputContext.FlushBuffersAsync()
-                .FollowOnSuccessWith(task => this.DisposeBatchWriterAndSetContentStreamRequestedState());
-        }
+        public abstract Task BatchOperationContentStreamRequestedAsync();
 #endif
 
         /// <summary>
         /// This method is called to notify that the content stream of a batch operation has been disposed.
         /// </summary>
-        void IODataBatchOperationListener.BatchOperationContentStreamDisposed()
-        {
-            Debug.Assert(this.CurrentOperationMessage != null, "Expected non-null operation message!");
-
-            this.SetState(BatchWriterState.OperationStreamDisposed);
-            this.CurrentOperationRequestMessage = null;
-            this.CurrentOperationResponseMessage = null;

```

Comment:
The downside of do this is that disposed will be spread in multipart reader and the base reader. Four lines of code is not big amount of duplicates but it makes the disposal logic all in one place for each class. Let me know how you think?